### PR TITLE
feat: Add floyd cycle-detection-algorithm to find duplicate number

### PR DESCRIPTION
**Pull request by Swastyy**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Relevant documentation/comments is changed or added
- [ ] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [ ] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C/pull/844"></a>

### CODE REVIEW

File: searching/floyd_cycle_detection_algorithm.c [hunk 1]
```diff
@@ -0,0 +1,66 @@
+/**
+ * @file floyd_cycle_detection_algorithm.c
```
Comment (Panquesito7) [inline searching/floyd_cycle_detection_algorithm.c:2]:
```suggestion
 * @file
```

File: searching/floyd_cycle_detection_algorithm.c [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+ * Given an array of integers containing 'n + 1' integers, where each
+ * integer is in the range [1, n] inclusive. If there is only one duplicate
+ * number in the input array, this algorithm returns the duplicate number in
+ * O(1) space and the time complexity is less than O(n^2) without modifying the
+ * original array, otherwise, it returns -1.
+ * @author [Swastika Gupta](https://github.com/Swastyy)
+ */
+#include <assert.h>
+#include <math.h>
+#include <stdio.h>
```
Comment (Panquesito7) [inline searching/floyd_cycle_detection_algorithm.c:15]: Please add a one-line description of what the library/header is for (**see the example below**).

```c
#include <stdio.h>      /// for IO operations
#include <assert.h>    /// for assert
```

File: searching/floyd_cycle_detection_algorithm.c [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+/**
+ * @brief The main function implements search algorithm
```
Comment (Panquesito7) [inline searching/floyd_cycle_detection_algorithm.c:18]:
```suggestion
 * @brief The main function implements the search algorithm
```

File: searching/floyd_cycle_detection_algorithm.c [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        tortoise = in_arr[tortoise];
+        hare = in_arr[hare];
+    }
+    return tortoise;
+}
+
+/**
+ * @brief Test implementation of the function
+ *
+ */
```
Comment (Panquesito7) [inline searching/floyd_cycle_detection_algorithm.c:47]:
```suggestion
/**
 * @brief Self-test implementations
 * @returns void
 */
```

File: searching/floyd_cycle_detection_algorithm.c [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+void test()
```
Comment (Panquesito7) [inline searching/floyd_cycle_detection_algorithm.c:48]:
```suggestion
static void test()
```

File: searching/floyd_cycle_detection_algorithm.c [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    size_t n = sizeof(arr) / sizeof(int);
+
+    printf("Test 1.... ");
+    int index = duplicateNumber(arr, n);
+    assert(index == 1);
+    printf("passed\n");
+}
+
+/**
+ * @brief Main function
```
Comment (Panquesito7) [inline searching/floyd_cycle_detection_algorithm.c:60]:
```suggestion
 * @brief Main function
 * @returns 0 on exit
```

File: searching/floyd_cycle_detection_algorithm.c [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+ */
+int main()
+{
+    test();
```
Comment (Panquesito7) [inline searching/floyd_cycle_detection_algorithm.c:64]:
```suggestion
    test();  // run self-test implementations
```

File: searching/floyd_cycle_detection_algorithm.c [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+ * @brief Implementation of [Floyd's Cycle
+ * Detection](https://en.wikipedia.org/wiki/Cycle_detection) algorithm
+ * @details
+ * Given an array of integers containing 'n + 1' integers, where each
+ * integer is in the range [1, n] inclusive. If there is only one duplicate
+ * number in the input array, this algorithm returns the duplicate number in
+ * O(1) space and the time complexity is less than O(n^2) without modifying the
+ * original array, otherwise, it returns -1.
+ * @author [Swastika Gupta](https://github.com/Swastyy)
+ */
```
Comment (Panquesito7) [inline searching/floyd_cycle_detection_algorithm.c:12]:
```suggestion
 */

```

File: searching/floyd_cycle_detection_algorithm.c [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+#include <stdio.h>
+
+/**
+ * @brief The main function implements search algorithm
+ * @tparam T type of array
+ * @param in_arr the input array
+ * @param n size of array
+ * @returns the duplicate number
+ */
+int duplicateNumber(const int *in_arr, size_t n)
```
Comment (Panquesito7) [inline searching/floyd_cycle_detection_algorithm.c:24]: Consider using `uint64_t` for non-negative values (or their appropriate size: `uint32_t`, `uint16_t`, `uint8_t`) or `int64_t` for negative values. **Requires adding the `inttypes.h` library.** Check other parts of the code ([reference](https://en.cppreference.com/w/c/types/integer)). 🙂

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Great work, @Swastyy! 😄
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/Swastyy/C/actions

File: searching/floyd_cycle_detection_algorithm.c [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+ * @brief Implementation of [Floyd's Cycle
+ * Detection](https://en.wikipedia.org/wiki/Cycle_detection) algorithm
+ * @details
+ * Given an array of integers containing 'n + 1' integers, where each
+ * integer is in the range [1, n] inclusive. If there is only one duplicate
+ * number in the input array, this algorithm returns the duplicate number in
+ * O(1) space and the time complexity is less than O(n^2) without modifying the
+ * original array, otherwise, it returns -1.
+ * @author [Swastika Gupta](https://github.com/Swastyy)
+ */
```
Comment (Panquesito7) [inline searching/floyd_cycle_detection_algorithm.c:12]:
```suggestion
 */
 
```

File: searching/floyd_cycle_detection_algorithm.c [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+ * @returns the duplicate number
+ */
+uint32_t duplicateNumber(const uint32_t *in_arr, size_t n)
+{
+    if (n == 0 || n == 1) {  // to find duplicate in an array its size should be atleast 2
+        return -1;
+    }
+    uint32_t tortoise = in_arr[0];  // variable tortoise is used for the longer
+                                    // jumps in the array
+    uint32_t hare = in_arr[0];  // variable hare is used for shorter jumps in the array
```
Comment (Panquesito7) [inline searching/floyd_cycle_detection_algorithm.c:31]:
```suggestion
    uint32_t tortoise = in_arr[0];  ///< variable tortoise is used for the longer
                                    ///< jumps in the array
    uint32_t hare = in_arr[0];  ///< variable hare is used for shorter jumps in the array
```

File: searching/floyd_cycle_detection_algorithm.c [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    assert(index == 1);
+    printf("passed\n");
+}
+
+/**
+ * @brief Main function
+ */
+int main()
+{
+    test();
```
Comment (Panquesito7) [inline searching/floyd_cycle_detection_algorithm.c:64]:
```suggestion
    test();  // run self-test implementations
```

File: searching/floyd_cycle_detection_algorithm.c [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+
+    printf("Test 1.... ");
+    uint32_t index = duplicateNumber(arr, n);
+    assert(index == 1);
+    printf("passed\n");
+}
+
+/**
+ * @brief Main function
+ */
```
Comment (Panquesito7) [inline searching/floyd_cycle_detection_algorithm.c:61]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: searching/floyd_cycle_detection_algorithm.c [hunk 14]
```diff
@@ -0,0 +1,67 @@
+/**
+ * @file
+ * @brief Implementation of [Floyd's Cycle
+ * Detection](https://en.wikipedia.org/wiki/Cycle_detection) algorithm
+ * @details
+ * Given an array of integers containing 'n + 1' integers, where each
```
Comment (Panquesito7) [inline searching/floyd_cycle_detection_algorithm.c:6]:
```suggestion
 * Given an array of integers containing `n + 1` integers, where each
```

File: searching/floyd_cycle_detection_algorithm.c [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+ 
+#include <assert.h>   /// for assert
+#include <inttypes.h> /// for uint32_t
+#include <stdio.h>    /// for IO operations
+
+/**
+ * @brief The main function implements the search algorithm
+ * @tparam T type of array
+ * @param in_arr the input array
+ * @param n size of array
```
Comment (Panquesito7) [inline searching/floyd_cycle_detection_algorithm.c:22]:
```suggestion
 * @param n size of the array
```

File: searching/floyd_cycle_detection_algorithm.c [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+ */
+static void test()
+{
+    uint32_t arr[] = {1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610};
+    size_t n = sizeof(arr) / sizeof(int);
+
+    printf("Test 1.... ");
+    uint32_t index = duplicateNumber(arr, n);
+    assert(index == 1);
+    printf("passed\n");
```
Comment (Panquesito7) [inline searching/floyd_cycle_detection_algorithm.c:57]: Please consider adding some documentation in this part of the code.

File: searching/floyd_cycle_detection_algorithm.c [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+/**
+ * @brief Main function
+ */
```
Comment (Panquesito7) [inline searching/floyd_cycle_detection_algorithm.c:62]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
```

File: searching/floyd_cycle_detection_algorithm.c [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+int main()
+{
+    test();
```
Comment (Panquesito7) [inline searching/floyd_cycle_detection_algorithm.c:65]:
```suggestion
    test();  // run self-test implementations
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: searching/floyd_cycle_detection_algorithm.c [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+    do {
+        tortoise = in_arr[tortoise];
+        hare = in_arr[in_arr[hare]];
+    } while (tortoise != hare);
+    tortoise = in_arr[0];
+    while (tortoise != hare) {
+        tortoise = in_arr[tortoise];
+        hare = in_arr[hare];
+    }
+    return tortoise;
```
Comment (Panquesito7) [inline searching/floyd_cycle_detection_algorithm.c:42]: Please add a bit of documentation of what this code does.

File: searching/floyd_cycle_detection_algorithm.c [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+    printf("passed\n");
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main()
+{
+    test();
```
Comment (Panquesito7) [inline searching/floyd_cycle_detection_algorithm.c:66]:
```suggestion
    test();  // run self-test implementations
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: searching/floyd_cycle_detection_algorithm.c [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test()
+{
+    uint32_t arr[] = {1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610}; // input array
+    size_t n = sizeof(arr) / sizeof(int);
+
+    printf("Test 1.... ");
```
Comment (Panquesito7) [inline searching/floyd_cycle_detection_algorithm.c:54]:
```suggestion
    printf("1st test... ");
```

**Review by Panquesito7 — COMMENTED** [review]
It's just to accept this suggestion, and you're done! 😄

**Comment by Swastyy** [comment]
> It's just to accept this suggestion, and you're done! 😄

Done

**Review by Panquesito7 — APPROVED** [review]
Amazing work, congratulations, @Swastyy! Thank you for your continued contributions to our community! 😄👍🎉

File: searching/floyd_cycle_detection_algorithm.c [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+/**
+ * @brief The main function implements the search algorithm
+ * @tparam T type of array
+ * @param in_arr the input array
+ * @param n size of the array
+ * @returns the duplicate number
+ */
+uint32_t duplicateNumber(const uint32_t *in_arr, size_t n)
+{
+    if (n == 0 || n == 1) {  // to find duplicate in an array its size should be atleast 2
```
Comment (mishraabhinn) [inline searching/floyd_cycle_detection_algorithm.c:27]:
```suggestion
    if (n <= 1) {  // to find duplicate in an array its size should be atleast 2
```

**Review by mishraabhinn — CHANGES_REQUESTED** [review]

**Review by Panquesito7 — APPROVED** [review]

**Review by mishraabhinn — APPROVED** [review]

**Pull request merged**